### ISSUE TITLE
Clicking on the headings in the outline moves the note to the wrong section

### ISSUE THREAD

**Issue by dsaperov** [GnoxNahte/obsidian-tab-panels#11]
### What happened?

Thank you so much for such a wonderful plugin. I usually put new cool plugins into the long grass, but this time I decided to implement it into my workflow the same day I saw it)

The problem I'm facing is that if the text inside the tab is long enough and contains a lot of headings, then clicking on one of the last (but not the last one) headings (subheadings) in the outline moves the note just below the place where this heading actually is.

I used markdown below and clicked on the subtitle "Condition monitoring". It was expected that the note would rewind down in such a way that the subtitle "Condition monitoring" itself would be at the top of the screen, but in my case, the top line that I see is "cholesterol and adjust the doses of medicines under the supervision of a doctor.", and 2 lines:
1. "Condition monitoring"
2. "When taking psyllium, it is important to regularly check the level of sugar,"
are above, off the screen.

![temp](https://github.com/user-attachments/assets/ef9c8bf4-cc9d-4a80-883f-f09d2a7094f4)

Could this be due to the fact that I'm not using the latest version of Obsidian or due to other plugins?

Thank you in advance!

### Sample Markdown

\`\`\`tabs
--- Tab 1
# Main conclusions and recommendations

### Lowering cholesterol

**Psyllium** helps to reduce general and "bad" cholesterol. The effect depends on the dosage (3-20 g per day) and the intake scheme. However, an incorrect reception time can not only negate the use, but also worsen the performance.

### Lowering blood pressure

**Psyllium** can reduce blood pressure by 2 mmHg, especially in people with initially high blood pressure. However, its effect is limited, and it does not affect the stiffness of the arteries.

### Constipation and hemorrhoids

**Psyllium** improves the frequency and quality of defecation, reduces the risk of hemorrhoids and intestinal cancer. It helps with both constipation and diarrhea.

### Diabetes

With the correct intake scheme, psyllium** reduces the level of glucose on an empty stomach and after meals, and also improves metformin intolerance.

### Heartburn

**Psyllium** reduces the symptoms of gastroesophageal reflux (heartburn) and reduces the frequency of relapses.

### Irritable bowel syndrome

Reduces the severity of symptoms by 16%, especially in combination with other types of cellulose, for example, ground flax.

### Ulcerative colitis

Reduces symptoms with regular intake.

### Gout

Reduces uric acid levels, which reduces the risk of gout.

### Weight loss

**Psyllium** increases the feeling of fullness, reduces the absorption of fat and promotes the redistribution of fat in the body, reducing its accumulation in the abdominal area.

### Wrong reception scheme

The reception time is critically important. For example, to reduce cholesterol **psyllium** should be taken with meals, and to reduce sugar — 20-30 minutes before meals.

### Contraindications

**Psyllium** is contraindicated in certain diseases (for example, intestinal obstruction) and requires caution when taken with drugs.

### Interaction with medicines

* Reduces the digestibility of levothyroxine (thyroid hormone) by 20%.
* Increases the bioavailability of metformin by 30%.
* May reduce the need for insulin.

It is recommended to take the drug 1 hour before or 4 hours after **psyllium**.

### Lead toxicity

**Psyllium** may contain lead, especially if it is grown in polluted regions ($\text{India}$, $\text{China}$). This requires choosing a product from safe countries ($\text{Germany}$, $\text{Canada}$, $\text{USA}$).

### Dosage

The optimal dose is 5-20 g per day, divided into 2-3 doses. The maximum safe dose is 30 g.

### Reception scheme

* To reduce cholesterol — during meals.
* To reduce sugar — 20-30 minutes before meals.
* For heartburn — an hour before meals.
* For constipation and hemorrhoids — 30 minutes before meals.

### Water

For each gram of **psyllium**, a minimum of 30 ml of water is needed to avoid suffocation or intestinal obstruction.

### Combination with other additives

To enhance the effect, it is recommended to combine **psyllium** with ground flax or other types of cellulose.

### Product Selection

Buy **psyllium** produced in low-pollution countries ($\text{Germany}$, $\text{Canada}$, $\text{USA}$). Avoid products from $\text{India}$ and $\text{China}$.

### A comprehensive approach

For maximum use, psyllium** should be combined with a change in diet, lifestyle, and other supplements.

### Condition monitoring

When taking **psyllium**, it is important to regularly check the level of sugar, cholesterol and adjust the doses of medications under the supervision of a doctor.

### Dosage table

The author has prepared a table with dosages and reception schemes for various diseases, which can be used as a guide.

# Conclusion

**Psyllium** is an effective supplement for the treatment of many diseases, but its use depends on the correct dosage, the time of intake and the quality of the product. The author emphasizes the importance of a scientific approach, complex treatment and individual selection of the reception scheme.
\`\`\`

### Obsidian Version

1.6.7

### OS

Windows

**Comment by GnoxNahte** [comment]
Hey, thanks for using and liking the plugin! Glad it's becoming part of your workflow

About the issue, when I tried it on my side, it just goes to the markdown... Not sure why it scrolls to the heading for you, maybe like you said, a different Obsidian version or another plugin is doing it. Either way, both aren't ideal.

I don't think I can solve it though... Outline is done by Obsidian and not the plugin. The plugin just adds to Obsidian's cache and Obsidian links it. 

In case you are curious about it (slightly technical), I added it to [HeadingCache](https://docs.obsidian.md/Reference/TypeScript+API/HeadingCache). If you see their documentation, the only related info I can add for linking is [position](https://docs.obsidian.md/Reference/TypeScript+API/CacheItem/position). This refers to where the heading is in the note, so I do not have much control over it, especially when comparing it with the rendered markdown in the tabs.

Though I might be wrong and there might be a solution for it. I'll search more on it tomorrow.

For reference, here's how it behaves on my end:
![TabPanels_HeadingsTest](https://github.com/user-attachments/assets/64c44e71-f400-4f96-a5fd-9dbbbc82df90)

**Comment by dsaperov** [comment]
Thanks for the quick response! On the gif that you attached the Editing mode is used, so I tried to do the same in Editing mode and everything worked out exactly the same way as yours. Could you please show how this operation works for you in Reading mode? Let me show how it works for me.
![gif](https://github.com/user-attachments/assets/bee1265d-9a93-4185-a986-d9a795f80731)

Is it an expected behaviour?

**Comment by GnoxNahte** [comment]
Oh right there's reading mode too. I tried it and I got the same result as you. But it's weird that it works for the top few headings but doesn't work for the last few headings. I'll test a bit more but as said before, I'm not sure if it can be fixed as the plugin just adds to the cache, and the Outline plugin handles navigating.

**Comment by dsaperov** [comment]
I have the following observations

## How navigation via Outline works normally

![gif](https://github.com/user-attachments/assets/1135b9a5-cad3-422c-b23f-878d2988bf2f)

You can notice the following:
1. The note will scroll exactly to the selected header (it will be located at the very top of the screen)
2. The header in the note is highlighted

## How navigation via Outline works inside tabs
### Upper headers

![gif1](https://github.com/user-attachments/assets/f757a6e9-4c32-424e-a9a9-e95ba6fe4509)

1. The note will scroll to the place above the selected header 
2. The header in the note is not highlighted

### Middle headers

![gif2](https://github.com/user-attachments/assets/f5594de0-008c-47d6-98bc-dcef22b8edfe)

1. If we click on each subheader in the outline, then somewhere in the middle of it we will find the one that, when clicked, will scroll the note exactly to the selected heading (but it is kinda coincidence)
2. The header in the note is still not highlighted

### Lower header

![gif3](https://github.com/user-attachments/assets/564b91d6-d982-48bf-a50b-5ad3efb9d1d7)

1. The note will scroll to the place below the selected header
2. The Tabs block is highlighted. In fact, when you select any of the headers in the Outline, the entire Tabs block is highlighted:

![gif4](https://github.com/user-attachments/assets/38751188-5d3f-4b8b-960b-dd37136209ca)

I freezed the page via `setTimeout(function(){debugger;}, 5000)` in the console to find out what is being flashed in the HTML and it turned out that it is really the whole block that is being flashed:

![temp](https://github.com/user-attachments/assets/3b8e704d-23ef-44d1-b0be-36ae9239da49)

Normally the selected header is being flashed:
![image](https://github.com/user-attachments/assets/602fa72a-30e6-45a2-94de-1254d45a7bca)

I understand that it is probably something we can't manage but may be observations above can shed light on why this is happening.

**Comment by GnoxNahte** [comment]
Wow, thanks for the detailed report! I wasn't expecting it haha. If you are in Obsidian's Discord, you can see [some of the testing](https://discord.com/channels/686053708261228577/840286264964022302/1318770544643936256) that I asked other devs too.

Was doing other stuff throughout the day but I was able to make a small prototype that I think might work? I'm still testing it though. Thanks for the report, I forgot that I should flash the highlight too. 

Clicking Heading 4:
![image](https://github.com/user-attachments/assets/16db653a-7e6f-4eda-92b0-7232b4699d28)

If you are curious how it's done, I registered a click event on the div. 
![image](https://github.com/user-attachments/assets/14b9638a-11f1-4c68-8a6b-0e0e1763c87f)
However, I don't see any documentation so I'm not sure if I'm doing it correctly. I just hope that this doesn't break some random thing somewhere else as this is sort of hijacking Obsidian's default behaviour.

Another problem I found (slightly unrelated), is if you link to the heading using something like [[File Name#Heading1]]. 
Since I have no control over this and I can't hijack anything, I think this really can't be solved, might put this in the readme.

**Comment by dsaperov** [comment]
No problem)

I'm glad you managed to fix it. I also hope it doesn't break anything. As for links like [[File Name#Heading 1]], even if there's nothing can be done, the fact that it will work in outline is already very good. Thanks for this!

**Comment by GnoxNahte** [comment]
Updated the plugin! Sorry it took some time. Let me know if it works and if there are any bugs.

This is in the release notes but I'll put it here too.
> [!WARNING]
> Note, these don't work and don't think it'll be fixed:
> - Scrolling on the page will highlight the wrong heading in Outline (Usually offsets by 1 heading)
> - It'll still position the page wrongly when using [[Note#Heading]]

**Comment by dsaperov** [comment]
Great! Thanks. Sorry, I don't have the opportunity to use obsidian rn, but next week I will be happy to update the plugin and let you know if something is wrong.

**Comment by dsaperov** [comment]
Hi! I just checked it and it works like a charm. Great, thank you!

**Comment by dsaperov** [comment]
Oh, I just noticed one thing. When I open a note and try to click on the header in the outline, I observe the behavior that I described here last week (the note is not being scrolled accurately to that header and the header in the note is not highlighted). But as soon as I switch to editing mode and back to reading view again, everything works fine. That is, it's as if the click event is registered only after switching to the editing view, while if I immediately open the note in the reading view, it doesn't work.

**Comment by dsaperov** [comment]
Noticed one more case of non-expected behaviour. After I perform the steps described in my previous post (in a note with Tab switch to Editing mode and back to Reading mode), if I switch to another already open note (or just open another note) and then switch back to a note with Tab, then again there will be the problem with navigating the headings. Sorry if I seem to be too focused on this plugin feature)

**Comment by dsaperov** [comment]
Thanks for fixing! I don't know why, after updating the plugin I face with a problem with navigating via headings in reading mode again (like in my first post). Now switching to editing mode --> reading mode doesn't help... Does it work correctly for you now?

**Comment by GnoxNahte** [comment]
It works for me though. 
![ezgif-7-aa7daaaa8f](https://github.com/user-attachments/assets/46f10245-270c-4ab2-974c-8d1caad7af65)

Do you mean like in this comment? https://github.com/GnoxNahte/obsidian-tab-panels/issues/11#issuecomment-2561775807

Try reloading Obsidian after updating? I didn't need to reload but it might fix random problems for you?

**Comment by dsaperov** [comment]
> Do you mean like in this comment?

Now I observe a behaviour like in the [first post)](https://github.com/GnoxNahte/obsidian-tab-panels/issues/11#issue-2745046907))

![gif](https://github.com/user-attachments/assets/fe5ab9f6-c2f0-4899-9904-45634a4dffac)

As you see there is an error in the console:
```
plugin:tab-panels:31 Outline heading length and headings in tab panels rendered markdown doesn't match up 
Outline Headings: (24) ['Main conclusions and recommendations', 'The benefits of psyllium', 'Lowering cholesterol', 'Lowering blood pressure', 'Constipation and hemorrhoids', 'Diabetes', 'Heartburn', 'Irritable bowel syndrome', 'Ulcerative colitis', 'Gout', 'Weight loss', 'Mistakes and risks', 'Wrong reception scheme', 'Contraindications', 'Interaction with medicines', 'Lead toxicity', 'Recommendations for admission', 'Dosage', 'Reception scheme', 'Water', 'Combination with other additives', 'Product Selection', 'Important remarks', 'A comprehensive approach'] 
Headings in plugin: (27) ['Main conclusions and recommendations', 'The benefits of psyllium', 'Lowering cholesterol', 'Lowering blood pressure', 'Constipation and hemorrhoids', 'Diabetes', 'Heartburn', 'Irritable bowel syndrome', 'Ulcerative colitis', 'Gout', 'Weight loss', 'Mistakes and risks', 'Wrong reception scheme', 'Contraindications', 'Interaction with medicines', 'Lead toxicity', 'Recommendations for admission', 'Dosage', 'Reception scheme', 'Water', 'Combination with other additives', 'Product Selection', 'Important remarks', 'A comprehensive approach', 'Condition monitoring', 'Dosage table', 'Conclusion']
```

It's like outline lacks some headings which present in the note. I tried to remove these headings from the note itself and it started to work fine. But when I add them back it doesn't work again. So I guess this bug can be reproduced using a text with a lot of headings (27 in my case).

> Btw sry for the bad gif quality. I know this is unrelated but if you are ok, can you share the software you used to create the gif?

No problem! I use [screen2gif](https://www.screentogif.com/) app. I've never used it before posting in this thread, so can't say anything else except that it's open source.

**Comment by GnoxNahte** [comment]
I found the problem! 

## Problem
I think Obsidian **doesn't** render all of the headings at once. 

To test if this is the same problem as you, 
1. expand Outline fully, so that it can fit all of the headings. 
2. reload the note
3. Try clicking the headings, it should link correctly

Then 
1. make the Outline as small as possible, maybe like your original gif make the scrollbar half of the outline.
<img width="161" alt="image" src="https://github.com/user-attachments/assets/ef7623d3-c16a-41f0-a65b-c98e211777e1" />

2. reload the note
3. Try clicking the headings, it should not link correctly.

## Solution?
Well as for whether it's fixable, I'm not sure... 
I can try to link the correct headings in the outline when the user loads the note. However, it has 2 problems:
- If the user scrolls in the Outline then it needs to update the links.
- The plugin might link to the wrong headings if there are duplicate headings (Not sure if anyone would do this but it'll be wrong anyway)

Anyway, I also found another problem. If there is a lot of text before the tab panels, it won't link correctly again as the tab panels aren't generated yet. I think this is Obsidian saving processing power, only running the code to convert the code block to the plugin's HTML when it's in view. Well I can't solve this as there's no way for me to link to something that's not generated.

I guess this is the problem with trying to do something that isn't really supported by Obsidian...

I think this might be quite confusing. My explanation can be better but if there's anything that you aren't sure feel free to ask. 

Thanks for the recommendation for the gif app! I'll try it when I have time. Nice that it's open source :)

**Comment by dsaperov** [comment]
Yeah, I managed to temporarily fix this by expanding Outline. And after making Outline small the bug returned just as you described.

I understand in general terms what the problem is, but my knowledge is not enough to offer any solution. I can only point out that at the time I wrote [this comment](https://github.com/GnoxNahte/obsidian-tab-panels/issues/11#issuecomment-2561775807), after the trick with switching to edit mode and back to read mode, it seems that navigation via headers in Outline started working normally (see the gif in that post). At least before opening another note. Maybe this can help somehow...

Anyway, if there is no solution to this problem, then so be it. I found a great use for the plugin with my notes on YouTube videos. It's like:
1st tab - emdedded video
2nd tab - its transcript
3rd tab - GPT's summary of the transcript. I used to have to break the page into 3 parts via corresponding headings, now thanks to the Tabs everything is much more compact and nice! So for me the problem with a lot of text before the tab panels is less actual, but summaries are usually quite long and have a lot of headings, so I guess this problem with Outline navigation will occur in such notes, but overall it's not critical and I'll continue to use the plugin anyway)
